calling-apis/managing-asynchronous-calls.md

```diff
@@ -0,0 +1,277 @@
+
+##### !end-question
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### Debugging Our Problem
+
+After debugging, Raffy should hone in on one specific line:
+
+```js
+const locations = findLocation(seattleCoordinates.seattleLat, seattleCoordinates.seattleLon);
+```
+
+Raffy found these two facts while debugging:
+
+- `seattleCoordinates.seattleLat` _and_ `seattleCoordinates.seattleLon` are `undefined` at the time of calling `findLocation`
+- The print statements in `findLocation` printed, meaning that `findLocation` was definitely called and got into the `catch` block
+
+After diagramming, thinking, researching, and rubber-ducking, Raffy comes to this conclusion:
+
+- `seattleCoordinates` has undefined values because it hasn't been assigned a value yet, and JavaScript gives `undefined` when accessing `seattleCoordinates.seattleLat` and `seattleCoordinates.seattleLon`
+- `seattleCoordinates` is assigned a value from `findLatitudeAndLongitude`'s return
+- `seattleCoordinates` wasn't assigned a value because `findLatitudeAndLongitude` didn't return yet
+- Something that changes the flow of control in our code is our asynchronous API calls
+
+### Refactoring: Utilize `then`
+
+How can Raffy ensure that `findLocation` executes after `findLatitudeAndLongitude` finishes?
+
+Raffy has a few options:
+
+- Re-order the code, so that the second API call happens **inside** the `then` chained to the first API call
+- Depending on our depth of knowledge in JavaScript, approach this from a different design pattern
+- Depending on different tools/libraries available, approach this from a different design pattern
+
+When working on a smaller JavaScript project, we can approach solving this problem by refactoring our work into the first `then` block.
+
+Raffy could [refactor his code to this](https://replit.com/@adacore/Managing-Asynchronous-Requests):
+
+<!-- prettier-ignore-start -->
+```js
+const findLocation = (latitude, longitude) => {
+  axios.get('https://us1.locationiq.com/v1/reverse.php',
+  {
+    params: {
+      key: YOUR_API_KEY,
+      format: 'json',
+      lat: latitude,
+      lon: longitude
+    }
+  })
+  .then( (response) => {
+    console.log('success in findLocation!', response.data);
+  })
+  .catch( (error) => {
+    console.log('error in findLocation!');
+  });
+}
+
+const findLocationFromLatAndLon = (query) => {
+  let latitude, longitude;
+  axios.get('https://us1.locationiq.com/v1/search.php',
+  {
+    params: {
+      key: YOUR_API_KEY,
+      q: 'Seattle, Washington, USA',
+      format: 'json'
+    }
+  })
+  .then( (response) => {
+    latitude = response.data[0].lat;
+    longitude = response.data[0].lon;
+    console.log('success in findLatitudeAndLongitude!', latitude, longitude);
+
+    findLocation(latitude, longitude);
+  })
+  .catch( (error) => {
+    console.log('error in findLatitudeAndLongitude!');
+  });
+}
+
+findLocationFromLatAndLon('Seattle, Washington, USA');
```

Comment:
Reordering this code makes this tougher to follow than it needs to be. Since the reference to `findLocation` is inside `findLatitudeAndLongitude`, the ordering doesn't matter.

For the renaming, that makes this function out of sync with the remaining questions, and also makes the code more complicated to understand. I agree that the previous name isn't great anymore since the function is doing other stuff, but I would like to see that mentioned in a section about drawbacks to this approach, and recommend that renaming is something the reader could do.

I also included the initial setup code for completeness.

These changes should also be made in replit.

```suggestion
const axios = require('axios');

const LOCATIONIQ_KEY = process.env['api_key'];

const findLatitudeAndLongitude = (query) => {
  let latitude, longitude;
  axios.get('https://us1.locationiq.com/v1/search.php',
  {
    params: {
      key: LOCATIONIQ_KEY,
      q: 'Seattle, Washington, USA',
      format: 'json'
    }
  })
  .then( (response) => {
    latitude = response.data[0].lat;
    longitude = response.data[0].lon;
    console.log('success in findLatitudeAndLongitude!', latitude, longitude);

    // make the next API call here!
    findLocation(latitude, longitude);
  })
  .catch( (error) => {
    console.log('error in findLatitudeAndLongitude!');
  });
}

const findLocation = (latitude, longitude) => {
  axios.get('https://us1.locationiq.com/v1/reverse.php',
  {
    params: {
      key: LOCATIONIQ_KEY,
      format: 'json',
      lat: latitude,
      lon: longitude
    }
  })
  .then( (response) => {
    console.log('success in findLocation!', response.data);
  })
  .catch( (error) => {
    console.log('error in findLocation!');
  });
}

findLatitudeAndLongitude('Seattle, Washington, USA');
```